rfcs/0076-refined-s3-source.md

```diff
@@ -0,0 +1,88 @@
+---
+feature: Refined S3 Source
+authors:
+  - "TabVersion"
+start_date: "2023/09/08"
+---
+
+# RFC: Refined S3 Source
+
+## Background and Motivation
+
+There has been two RFCs ([RFC: Refine S3 Source: List Objects in A Lazy Way](https://github.com/risingwavelabs/rfcs/pull/74) 
+and [RFC: S3 Source with SQS](https://github.com/risingwavelabs/rfcs/pull/22)) talking about S3 source but neither of them 
+fits the need of RisingWave. This RFC is to combine the pros of both RFCs and make S3 source elegant.
+
+The main problems are:
+
+* There can be millions of files in s3, but we cannot store all items in Meta's memory or ETCD.
+* Pushing down the file assignment from Meta to Compute Node can be burdensome, and it makes barrier stuck if it carries 
+  too many files.
+
+# Solution
+
+The intuitive solution is to use state store to store the file list, and it is kind of a good idea.
+But it is not easy for Meta to access Hummock, the storage engine, and even if it could write the file list down to state
+store, it is even harder to align the data with barrier, which means it can be dangerous for source executors to use the list.
+
+## List and Fetch
+
+So I am proposing a new solution, which is to use two separate executors to complete the job.
+Let's say they are `list` and `fetch`, aligning the name in [RFC: S3 Source with SQS](https://github.com/risingwavelabs/rfcs/pull/22), 
+and they do the same thing.
+
+* `List` is unique in one source plan. It is responsible for keep listing all files and push the filenames to the downstream.
+* `Fetch` is responsible for fetching the file from S3 and push the file to the downstream, as well as write the received
+  filenames to Hummock.
+
+## How It Works
+
+![img](../assets/refined-s3-source.svg)
+
+### List Part
+
+The `list` cannot push all filenames to the `fetch` part every time. Letting the shuffle operator distribute millions of
+rows every minute is an excessive and unnecessary burden on the system.
+If we do this, the fetch operator must perform `point get` operation on each row to deduplicate, which is also unacceptable for hummock.
+
+The `list` executor will keep listing the files in S3 bucket and try its best to de-duplicate the files (but no guarantee).
+It will record the timestamp when doing the last listing and filter out the files that are created before the timestamp.
+And we can utilize the idea of [RFC: Refine S3 Source: List Objects in A Lazy Way](https://github.com/risingwavelabs/rfcs/pull/74),
+not to list all pages at once, but to list the pages one by one, preventing overload downstream.
+
+`list` executor will keep the max page `p` number as well as the last list timestamp `ts` in its state store.
```

Comment:
May I ask how can we find the files being deleted?

Reply:
No. `list` will only handle the new added ones. If `fetch` meets an invalid file path, it will be discarded.